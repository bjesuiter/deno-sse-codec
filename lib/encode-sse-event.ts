import { SSEEvent } from "./types/sse-event.ts";
import { IllegalArgumentError } from "./types/illegal-argument-error.ts";

/**
 * Encodes a SSEEvent into a string which could be sent over the wire like this.
 * @param event
 */
export function encodeSSEEvent(
  { eventName, data, id, retry }: SSEEvent,
): string {
  if (!data && !eventName && !id) {
    throw new IllegalArgumentError(`Input SSEEvent does not contain any data! 
    You need minimum one of SSEEvent.name, SSEEvent.data or SSEEvent.id`);
  }

  let sseString = "";

  if (eventName) {
    sseString += `event: ${eventName} \n`;
  }

  if (data) {
    sseString += `data: ${data} \n`;
  }

  if (id) {
    sseString += `id: ${id} \n`;
  }

  if (retry) {
    sseString += `retry: ${retry} \n`;
  }

  // add an empty line after the completed event
  sseString += `\n`;

  return sseString;
}
