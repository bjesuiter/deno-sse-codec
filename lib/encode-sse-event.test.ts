import {
  assertSnapshot,
  assertThrows,
  describe,
  it,
} from "../dependencies/_testing.std.ts";
import { encodeSSEEvent } from "./encode-sse-event.ts";

describe(`encode-sse-event`, () => {
  it(`should throw on empty SSEEvent argument`, () => {
    assertThrows(() => encodeSSEEvent({}));
  });
  it(`should throw on a 'retry'-property-only SSEEvent argument`, () => {
    assertThrows(() => encodeSSEEvent({ retry: 5000 }));
  });

  it(`should encode single properties`, async (t) => {
    await assertSnapshot(t, encodeSSEEvent({ id: "UID5346324874238475" }));
    await assertSnapshot(t, encodeSSEEvent({ name: "custom-event" }));
    await assertSnapshot(
      t,
      encodeSSEEvent({ data: "Some simple string data" }),
    );
    await assertSnapshot(
      t,
      encodeSSEEvent({ name: "custom-event", retry: 5000 }),
    );
  });

  it(`should encode a complete SSEEvent`, async (t) => {
    await assertSnapshot(
      t,
      encodeSSEEvent({
        name: "custom-event",
        data: "Some simple string data",
        id: "UID5346324874238475",
        retry: 5000,
      }),
    );
  });
});
