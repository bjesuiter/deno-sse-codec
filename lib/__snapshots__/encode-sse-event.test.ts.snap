export const snapshot = {};

snapshot[`encode-sse-event > should encode single properties 1`] = `
"id: UID5346324874238475 

"
`;

snapshot[`encode-sse-event > should encode single properties 2`] = `
"event: custom-event 

"
`;

snapshot[`encode-sse-event > should encode single properties 3`] = `
"data: Some simple string data 

"
`;

snapshot[`encode-sse-event > should encode single properties 4`] = `
"event: custom-event 
retry: 5000 

"
`;

snapshot[`encode-sse-event > should encode a complete SSEEvent 1`] = `
"event: custom-event 
data: Some simple string data 
id: UID5346324874238475 
retry: 5000 

"
`;
